Libcloud 386 GCE Load Balancer Support

wrigri opened the pull request:
This adds Load Balancing support for Google Compute Engine.

Since Load Balancer support for GCE is directly part of the GCE API, all of the API calls were added to the existing compute driver.  Then, wrapper methods were added to the Load Balancer driver to make it compatible with the Libcloud Load Balancer API.

In addition, there are some bug fixes and minor improvements.

Note that I'm still working on the test for the "wrapper" methods in the LoadBalancer driver, but tests have been added for all of the actual API calls that were added to the Compute driver.

wrigri committed:
Initial commit of Load Balancer support for GCE.
Since the GCE Load Balancer is a part of the GCE API, the API calls have all
been added to the compute driver, and the LB driver wraps those methods.

wrigri committed:
Add Load Balancing demo code, plus more bug fixes.
Persistent disk support has been added to multiple disk creation, but it still
needs some work since the disks are created serially instead of in parallel.

wrigri committed:
Adding modified files that got left out of last commit.

wrigri committed:
Fix issue with target pool containing non-existent nodes.
Also some PEP8 and misc fixes

wrigri committed:
Updates to GCE compute tests for Load Balancer features.

wrigri committed:
A couple of Python 3 fixes.

wrigri committed:
Fix Cloudstack entries that got accidentially overwritten.

wrigri committed:
Create a common way for dealing with 404/Resource not found errors.
Also, fix for an issue that isn't setting defalut values in healthchecks.

wrigri committed:
Add 'Resource in Use' Error.
(plus a pep8 fix)

Kami commented:
Thanks and sorry for the delay. This pull request is slightly larger then usual so I will take a more in depth look sometime in the next couple of days.

We've recently bootstrap new Sphinx documentation so it would be great if you can add some usage example and docs about the GCE drivers there - https://github.com/apache/libcloud/tree/trunk/docs

wrigri commented:
No problem.  As you can see, I've been making some additional minor fixes as a colleague of mine has been using the GCE driver.  I'm happy to contribute some documentation, so I'll start looking at that.

wrigri committed:
There is more than one "RESOURCE_IN_USE" error, so do a partial match.

wrigri committed:
Unify errors so that they all keep the 'code' attribute, even if known.

Kami commented on demos/gce_lb_demo.py:
```diff
@@ -0,0 +1,293 @@
+try:
+    import secrets
+except ImportError:
+    secrets = None
```
Hm, shouldn't we exit here if `secrets.py` file doesn't exist and print a friendly message which tells the user how to update the secrets file?

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -499,6 +711,47 @@ def ex_list_firewalls(self):
+        # Use provided region or default region
+        region = region or self.region
+        # Setting region to 'all' overrides the default region
+        if region == 'all':
```
I noticed this pattern repeated a couple of times.

Is it possible to refactor it in a separate function and re-use this function?

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1006,6 +1440,58 @@ def ex_create_multiple_nodes(self, base_name, size, image, number,
+
+        if healthchecks:
+            if not hasattr(healthchecks[0], 'name'):
+                hc_list = [self.ex_get_healthcheck(h).extra['selfLink'] for h
```
I also noticed this repeated a couple of times - it would be good to add a note to documentation which says something along the lines of "Some functions allows you to pass in either a list of objects or a list strings. In some cases passing in a list of strings will result in additional <list length> HTTP requests".

wrigri replied:
I added some additional information to the docstring for the GCE Driver.  Let me know if that makes sense to you or not.

Kami replied:
Looks good.

Need to make sure we also don't forget to add this to Sphinx documentation once it's available.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1051,7 +1539,42 @@ def create_volume(self, size, name, location=None, image=None,
+        response = self.connection.async_request(request, method='PUT',
+                                                 data=hc_data).object
+        if 'error' in response:
+            self._categorize_error(response['error'])
```
I also noticed this pattern repeated all over.

Can we use the response status code and centralize the error handling in the `Response` class so we don't need to repeat this if check in many function bodies?

[Here](https://github.com/apache/libcloud/blob/trunk/libcloud/dns/drivers/zerigo.py#L66-L95) is an example of how we do that in the Zerigo driver.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1086,6 +1609,138 @@ def ex_update_firewall(self, firewall):
+            index = None
+            for i, nd in enumerate(targetpool.nodes):
+                if nd.name == node.name:
+                    index = i
```
Can we break here on the first match?

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1407,17 +2120,36 @@ def ex_get_address(self, name, region=None):
-        request = '/regions/%s/addresses/%s' % (address_region, name)
+        region = region or self.region or self._find_zone(name, 'addresses',
+                                                          region=True)
+        if not region:
```
Sam here - can we refactor the next couple of lines in a separate function?

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1407,17 +2120,36 @@ def ex_get_address(self, name, region=None):
+                                                          region=True)
+        if not region:
+            raise ResourceNotFoundError(
+                'Address %s not found in any region.' % name, 404)
```
I prefer to always use parenthesis around the argument(s) when doing string formatting. I've seen missing parenthesis cause explosions multiple times - people add another argument, but forgot to add parenthesis.

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1559,7 +2373,12 @@ def ex_get_zone(self, name):
+        # If zone is not found, the response is html and not json, so it fails
+        # parsing
+        except MalformedResponseError:
+            return None
```
(off topic)

Google, that's kinda nasty... But then again, I've also seen a lot of worse things when working with other provider APIs :P

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1767,10 +2674,45 @@ def _to_storage_volume(self, volume):
+                            in targetpool.get('healthChecks', [])]
+        node_list = []
+        for n in targetpool.get('instances', []):
+            node_path = n.split('/')
```
It would be good to split this into a separate function which returns a dict - e.g. `get_components_from_node_path.`

Kami commented on libcloud/test/compute/test_gce.py:
```diff
@@ -96,13 +109,27 @@ def test_ex_list_addresses(self):
         self.assertEqual(address_list_uc1[0].name, 'libcloud-demo-address')
-        #self.assertEqual(address_list_all[0].name, 'lcaddress')
+
+    def test_ex_list_healthchecks(self):
```
Side note about the tests - a bunch of compute tests randomly fail under Python 3.3 in trunk.

I think that is because some of the tests rely on the dictionary being ordered. This is generally a bad practice and won't work in Python 3.3 where hash randomization has been enabled by default (http://docs.python.org/3/whatsnew/3.3.html).

For now I commented some of those tests out, but it would be great if you could change them to not rely on the dictionary ordering.

wrigri commented:
Thanks for the comments, I should be able to address most of them today. A couple of notes:
- Like you, I wasn't happy with the way that various exceptions were being handled, so I re-wrote the exception handling in a separate branch last week which addresses several of the issues that you noted (including that MalformedResponseError for missing zones that occurred because one of the zones was in a weird state the week before last, and moving it into the Response class).  I'll merge that branch in today (but squash the commits).
- Thanks for the heads-up about the dictionary ordering thing.  I'm actually already aware of that, since I've used previous versions of Python where the randomization has been turned on.  I'm surprised I let any of those slip through, but I'll track them down and get them fixed.

Kami commented:
Sounds good. Thanks!

wrigri committed:
Merge in re-write of Exception handling from the ExceptionUpdate branch

wrigri committed:
Fixes based on pull request feedback.
Also, added some additional information in the user-agent.

wrigri commented:
Do you have some examples of the tests that are failing?  I can't seem to reproduce the failures when I turn on hash randomization (and I was able to put in a temporary test that _would_ fail intermittently with the randomization turned on, to verify that it was actually randomized).

wrigri commented:
Nevermind.  I see what you are referring to on the tests now.  (I hadn't seen that you had committed the change, so I didn't notice when I merged those changes into my branch.)  Anyway, i'll add back some extra tests to replace the removed ones.

wrigri committed:
Fix tests and replace old tests for aggregated results.

Kami commented on libcloud/common/google.py:
```diff
@@ -101,10 +103,121 @@ def __repr__(self):
+            body = self.body
+            json_error = True
+
+        if self.status in [httplib.OK, httplib.CREATED, httplib.ACCEPTED]:
```
Slightly confused - does Google always return those codes, even on error?

wrigri replied:
Correct.  In many cases, there can be "errors" that still return a 200 code (which was why I was previously looking for "error" in the JSON response, and then calling that _categorize_error() method.  Except for the 404, most "expected" errors do not return http error codes.

Kami replied:
I see. Yeah, that's unfortunate and means we can't really split it up and move most of the error parsing to `parse_error`.

You should also bug your colleagues to use 4xx error code for error conditions and make everyone's life easier :)

Kami commented on libcloud/common/google.py:
```diff
@@ -101,10 +103,121 @@ def __repr__(self):
+        message = err.get('message')
+        return (code, message)
+
+    def parse_body(self):
```
This looks way better, but a the "libcloud way" would be to do the following:
- Modify `success` method to only return `True` if the status code indicates success
- If `success` returns `False`, `parse_error` will be called where you can parse error from the body and raise appropriate exception

This approach relies on provider returning appropriate status codes on error response though. Looking at the code bellow, this might not be the case though.

wrigri replied:
I could probably split it up a bit, and add the 404 and other errors to parse_error, but I would still need to leave the ResourceInUse/QuotaExceeded/etc. error parsing in parse_body.

wrigri committed:
Fix an error that comes up region/zone and self.region/self.zone are None

Kami commented on demos/gce_demo.py:
```diff
@@ -41,7 +41,13 @@
+          'Additional information about setting these values can be found '
+          'in the docstring for:\n'
+          'libcloud/common/google.py\n')
+    sys.exit()
```
Minor thing - exiting with a non-zero code would be better.

Kami commented on demos/gce_lb_demo.py:
```diff
@@ -0,0 +1,299 @@
+          'Additional information about setting these values can be found '
+          'in the docstring for:\n'
+          'libcloud/common/google.py\n')
+    sys.exit()
```
Same here.

Kami commented on demos/gce_lb_demo.py:
```diff
@@ -0,0 +1,299 @@
+
+# ==== HELPER FUNCTIONS ====
+def get_gce_driver():
+    driver = get_driver(Provider.GCE)(*args, datacenter=DATACENTER, **kwargs)
```
Does this pass under Python 2.5? IIRC you can't mix kwargs with arg and kwarg expansion.

Kami replied:
Yeah, just checked it:

``` python

Python 2.5.6 (r256:88840, Aug 21 2013, 17:00:00) 
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> def foo(a, b, c): pass

>>> foo(*(1,2), b='foo', **{'c': ''})
  File "<stdin>", line 1
    foo(*(1,2), b='foo', **{'c': ''})
                ^
SyntaxError: invalid syntax
```

Solution would be to only use arg and kwarg expansion. So doing something like:

``` python
kwargs = kwargs.copy()
kwargs['datacenter'] = DATACENTER
...
```

Kami commented on demos/gce_lb_demo.py:
```diff
@@ -0,0 +1,299 @@
+    if gce_driver:
+        driver = get_driver_lb(Provider_lb.GCE)(gce_driver=gce_driver)
+    else:
+        driver = get_driver_lb(Provider_lb.GCE)(*args, datacenter=DATACENTER,
```
Same here (with regards to mixing arg expansion + kwargs).

Kami commented on libcloud/compute/drivers/gce.py:
```diff
@@ -1353,7 +2004,14 @@ def ex_destroy_multiple_nodes(self, nodelist, ignore_errors=True,
+            except:
+                e = sys.exc_info()[1]
+                response = None
+                if not ignore_errors:
```
I won't if we could refactor this `ignore_errors` functionality in a way so it can be reused.

Kami commented:
I've looked over the pull request again.

It's close - there are some minor issues left + missing load balancer wrapper tests.

wrigri committed:
Unittests for Loadbalancer wrapper plus minor fixes based on feedback.

wrigri committed:
Missed one of the datacenter=DATACENTER keywords.

wrigri commented:
OK, I think that I've addressed all of the comments.  Let me know if there are any other final things that need to be done before we can get this merged in.

wrigri committed:
Fix up docstring and add license

Kami commented:
@wrigri Thanks and sorry for the delay.

We don't receive notifications when you add a comment so next time please mention my nickname so I will receive a notification and respond faster.

Anyway, please squash all the commits and attach latest patch to the JIRA ticket and I will get it merged.

wrigri commented:
@Kami Thanks for taking a look.  I have attached a patch to the ticket.  (And sorry for not including your nickname, I'll make sure I do that in the future.)

Kami commented:
Forgot to add a comment - this pull request has been merged and it can be closed now.

Thanks again!

wrigri commented:
Closing pull request.
